File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3108_minimum-cost-walk-in-weighted-graph/solutions/python3/solution.py
from collections import defaultdict
from heapq import heappush, heappop
import sys

class Solution:
    def minimumCost(self, n: int, edges: List[List[int]], query: List[List[int]]) -> List[int]:
        # Create adjacency list for the graph
        graph = defaultdict(list)
        for u, v, w in edges:
            graph[u].append((v, w))
            graph[v].append((u, w))
        
        def minimum_and_cost(si: int, ti: int) -> int:
            # Min-heap for prioritizing which node to explore next
            # Initialize heap with the starting node and "infinity" bitwise AND (all bits set)
            heap = [(0, si)]
            # Dist to keep track of the minimum AND cost from si to all other nodes
            dist = {i: sys.maxsize for i in range(n)}
            dist[si] = -1  # Bitwise AND identity is all bits set (so starting with a not possible value)

            while heap:
                current_cost, node = heappop(heap)
                
                # If the current node is the target and we have not popped it before
                if node == ti:
                    return current_cost if current_cost != sys.maxsize else -1

                # Explore all adjacent nodes
                for neighbor, weight in graph[node]:
                    # Calculate new cost with AND operation
                    new_cost = current_cost & weight if current_cost != 0 else weight
                    # Only consider this new path if it offers a smaller AND cost
                    if new_cost < dist[neighbor]:
                        dist[neighbor] = new_cost
                        heappush(heap, (new_cost, neighbor))

            # If no path was found to target, return -1
            return -1 if dist[ti] == sys.maxsize else dist[ti]
        
        # Process each query and determine the minimum AND cost for each
        result = []
        for si, ti in query:
            result.append(minimum_and_cost(si, ti))
        
        return result

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3108_minimum-cost-walk-in-weighted-graph/solutions/cpp/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <queue>
#include <limits.h>

using namespace std;

class Solution {
public:
    vector<int> minimumCost(int n, vector<vector<int>>& edges, vector<vector<int>>& queries) {
        vector<unordered_map<int, int>> graph(n);

        // Build the graph with the minimum AND-ed weight on each edge
        for (const auto& e : edges) {
            int u = e[0], v = e[1], w = e[2];
            if (!graph[u].count(v) || (graph[u][v] & w) < graph[u][v]) {
                graph[u][v] = w;
                graph[v][u] = w;
            }
        }

        vector<int> results;
        for (const auto& q : queries) {
            int start = q[0], end = q[1];
            results.push_back(bfs(graph, start, end, n));
        }

        return results;
    }

private:
    int bfs(const vector<unordered_map<int, int>>& graph, int start, int end, int n) {
        vector<int> min_cost(n, INT_MAX); // minimum AND cost to reach each node
        queue<int> q;
        q.push(start);
        min_cost[start] = ~0; // start with all bits set (AND identity)

        while (!q.empty()) {
            int node = q.front();
            q.pop();

            for (const auto& p : graph[node]) {
                int neighbor = p.first;
                int edge_cost = p.second;
                int new_cost = min_cost[node] & edge_cost;

                if (new_cost < min_cost[neighbor]) {
                    min_cost[neighbor] = new_cost;
                    q.push(neighbor);
                }
            }
        }

        return min_cost[end] == INT_MAX ? -1 : min_cost[end];
    }
};
